guppy/src/graph/cargo.rs

```diff
@@ -351,6 +351,98 @@ impl<'a> CargoSetBuildState<'a> {
         }
     }
 
+    fn new_v2(self, query: FeatureQuery<'_>) -> CargoSet {
```

Comment:
The core algorithm starts here. Note the XXX marks about whether some cases are bugs in upstream cargo.